draft-barnes-acme.md

```diff
@@ -1603,87 +1604,90 @@ failed.
+accessing the created TLS server and verifying a particular challenge
+certificate is presented.
 
 type (required, string):
 : The string "dvsni"
 
-r (required, string):
-: A random 32-byte octet, Base64-encoded
+token (required, string):
+: A random 16-byte octet string, hex-encoded
+
+~~~~~~~~~~
+{
+  "type": "dvsni",
+  "token": "a82d5ff8ef740d12881f6d3c2277ab2e",
+}
+~~~~~~~~~~
+
+In response to the challenge, the client uses its account private key to sign a
+JWS over a JSON object describing the challenge.  The validation object covered
+by the signature MUST have the following fields and no others:
+
+type (required, string):
+: The string "dvsni"
 
-nonce (required, string):
-: A random 16-byte octet string, hex-encoded (so that it can be used as a DNS label)
+token (required, string):
+: A random 16-byte octet string, hex-encoded
 
 ~~~~~~~~~~
 {
   "type": "dvsni",
-  "r": "Tyq0La3slT7tqQ0wlOiXnCY2vyez7Zo5blgPJ1xt5xI",
-  "nonce": "a82d5ff8ef740d12881f6d3c2277ab2e"
+  "token": "a82d5ff8ef740d12881f6d3c2277ab2e",
 }
 ~~~~~~~~~~
 
-The client responds to this Challenge by configuring a TLS server on port 443 of
-a server with the domain name being validated:
+The client serializes the validation object to UTF-8, then uses its account
+private key to sign a JWS with the serialized JSON object as its payload. This
+JWS is NOT REQUIRED to have the "nonce" header parameter.
 
-1. Decode the server's random value R
-2. Generate a random 32-byte octet string S
-3. Compute Z = SHA-256(R &#124;&#124; S) (where &#124;&#124; denotes
-   concatenation of octet strings)
-4. Generate a self-signed certificate with a subjectAltName extension containing
-   two dNSName values:
-  1. The domain name being validated
-  2. A name formed by hex-encoding Z and appending the suffix ".acme.invalid"
-5. Compute a nonce domain name by appending the suffix ".acme.invalid" to the
-   nonce provided by the server.
-6. Configure the TLS server such that when a client presents the nonce domain
-   name in the SNI field, the server presents the generated certificate.
+The client will generate a self-signed certificate with the subject's
+organizationName field set to the "signature" value from the JWS, i.e., the
+base64-encoded signature value and a subjectAlternativeName extension containing
+a single dNSName of "\<token\>.acme.invalid".  The client will then configure
+the TLS server at the domain such that when a handshake is initiated with the
+Server Name Indication extension set to "\<token\>.acme.invalid", the generated
+test certificate is presented.
 
-The client's response provides its random value S:
+The response to the DVSNI challenge provides the validation JWS to the server.
 
 type (required, string):
 : The string "dvsni"
 
-s (required, string):
-: A random 32-byte secret octet string, Base64-encoded
+validation (required, string):
+: Base64-encoded SHA-256 hash of the DER-encoded challenge certificate
```

Comment:
This does not match what you've got below.  It doesn't seem like a bad idea to have the cert hash in the response, but you need to provide the JWS itself, either in the response or in the cert (probably the former).